File: content_generation/translate.py
from langchain.prompts.chat import (
    ChatPromptTemplate,
    HumanMessagePromptTemplate,
    SystemMessagePromptTemplate,
)

from langchain_openai import ChatOpenAI
from langchain.schema import Document
from langchain_community.document_transformers import GoogleTranslateTransformer

import os
from dotenv import load_dotenv

load_dotenv()

def convert_language_to_code(language: str):
    language_codes = {
        "english": "en",
        "spanish": "es",
        "french": "fr",
        "german": "de",
        "italian": "it",
        "portuguese": "pt",
        "dutch": "nl",
        "russian": "ru",
        "japanese": "ja",
        "chinese": "zh",
        "korean": "ko",
        "arabic": "ar",
        "turkish": "tr",
        "hindi": "hi",
        "indonesian": "id",
        "vietnamese": "vi",
        "thai": "th",
        "greek": "el",
        "hebrew": "he",
        "swedish": "sv",
        "danish": "da",
        "norwegian": "no",
        "finnish": "fi",
        "czech": "cs",
        "polish": "pl",
        "hungarian": "hu",
        "romanian": "ro",
        "ukrainian": "uk",
        "catalan": "ca",
        "vietnamese": "vi",
        "tagalog": "tl",
        "malay": "ms",
        "swahili": "sw",
        "afrikaans": "af",
        "esperanto": "eo",
        "latin": "la",
        "welsh": "cy",
        "scots gaelic": "gd",
        "irish": "ga",
        "icelandic": "is",
        "farsi": "fa",
        "serbian": "sr",
        "croatian": "hr",
        "bosnian": "bs",
        "slovak": "sk",
        "slovenian": "sl",
        "estonian": "et",
        "latvian": "lv",
        "lithuanian": "lt",
        "maltese": "mt",
        "albanian": "sq",
        "macedonian": "mk",
        "kurdish": "ku",
        "armenian": "hy",
        "georgian": "ka",
        "basque": "eu",
        "galician": "gl",
        "catalan": "ca",
        "corsican": "co",
    }
    return language_codes.get(language.lower(), "en")


def translate_text(text: str, target_language: str):
    documents = [Document(page_content=text)]
    translator = GoogleTranslateTransformer(project_id=os.environ["GOOGLE_CLOUD_PROJECT_ID"])
    docs = translator.transform_documents(documents, target_language_code=convert_language_to_code(target_language))
    return {
        "translate": next((doc.page_content for doc in docs if doc.page_content), "No translation available")
    }

File: src/endpoints/translate.ts
import {
    OpenAPIRoute,
    OpenAPIRouteSchema,
} from "@cloudflare/itty-router-openapi";
import { Ai } from '@cloudflare/ai'
import {TranslatePrompt} from "../types";
import type { AiEnv } from "../types";

const languageCodes = ["english",
    "spanish",
    "french",
    "german",
    "italian",
    "portuguese",
    "dutch",
    "russian",
    "japanese",
    "chinese",
    "korean",
    "arabic",
    "turkish",
    "hindi",
    "indonesian",
    "vietnamese",
    "thai",
    "greek",
    "hebrew",
    "swedish",
    "danish",
    "norwegian",
    "finnish",
    "czech",
    "polish",
    "hungarian",
    "romanian",
    "ukrainian",
    "catalan",
    "vietnamese",
    "tagalog",
    "malay",
    "swahili",
    "afrikaans",
    "esperanto",
    "latin",
    "welsh",
    "scots gaelic",
    "irish",
    "icelandic",
    "farsi",
    "serbian",
    "croatian",
    "bosnian",
    "slovak",
    "slovenian",
    "estonian",
    "latvian",
    "lithuanian",
    "maltese",
    "albanian",
    "macedonian",
    "kurdish",
    "armenian",
    "georgian",
    "basque",
    "galician",
    "catalan",
    "corsican"]

export class Translate extends OpenAPIRoute {
    static schema: OpenAPIRouteSchema = {
        tags: ["Translate"],
        summary: "Translate a text",
        requestBody: TranslatePrompt,
        responses: {
            "200": {
                description: "Returns the translated text",
                schema: {},
            },
        },
    };

    async handle(
        request: Request,
        env: AiEnv,
        context: any,
        data: Record<string, any>
    ) {
        // Retrieve the validated request body
        const ai = new Ai(env.AI);

        const inputs = {
            text: data.body.text,
            source_lang: data.body.sourceLanguage,
            target_lang: data.body.targetLanguage,
        }

        if (!languageCodes.includes(inputs.source_lang.toLowerCase())) {
            return Response.json({
                text: "Invalid source language"
            })
        }

        if (!languageCodes.includes(inputs.target_lang.toLowerCase())) {
            return Response.json({
                text: "Invalid target language"
            })
        }

        const response = await ai.run(
            "@cf/meta/m2m100-1.2b",
            inputs
        );

        return Response.json({
            text: response.translated_text
        })
    }
}